src/librbd/api/Config.cc

```diff
@@ -0,0 +1,366 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/api/Config.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Utils.h"
+#include "librbd/api/PoolMetadata.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Config: " << __func__ << ": "
+
+namespace librbd {
+namespace api {
+
+namespace {
+
+const uint32_t MAX_KEYS = 64;
+
+typedef std::map<std::string, std::pair<std::string, config_source_t>> Parent;
+
+struct Options : Parent {
+  Options(bool image_apply_only_options) :
+    Parent{
```

Comment:
It would be great if the existing ```metadata_set``` and the new ```pool_metadata_set``` APIs would validate any ```conf_XYZ``` keys against the valid list.

Reply:
Personally I like an idea that with "metadata" methods you can set any keys without any restrictions and use "config" methods if you want validation, but again I don't have a strong opinion here.

Reply:
I was suggesting that you drop all the new config API methods and just use the existing metadata APIs for the CLI's config actions.